BinaryEncoding.md

```diff
@@ -317,18 +317,20 @@ It is legal to have several entries with the same type.
 | Name | Opcode | Immediate | Description |
 | ---- | ---- | ---- | ---- |
 | `nop` | `0x00` | | no operation |
-| `block` | `0x01` | count = `varuint32` | a sequence of expressions, the last of which yields a value |
-| `loop` | `0x02` | count = `varuint32` | a block which can also form control flow loops |
-| `if` | `0x03` | | high-level one-armed if |
-| `if_else` | `0x04` | | high-level two-armed if |
+| `block` | `0x01` |  | begin a sequence of expressions, the last of which yields a value |
+| `loop` | `0x02` |  | begin a block which can also form control flow loops |
+| `if` | `0x03` | | begin if expression |
+| `else` | `0x04` | | begin else expression of if |
 | `select` | `0x05` | | select one of two values based on condition |
 | `br` | `0x06` | relative_depth = `varuint32` | break that targets a outer nested block |
 | `br_if` | `0x07` | relative_depth = `varuint32` | conditional break that targets a outer nested block |
 | `br_table` | `0x08` | see below | branch table control flow construct |
-| `return` | `0x14` | | return zero or one value from this function |
-| `unreachable` | `0x15` | | trap immediately |
+| `return` | `0x09` | | return zero or one value from this function |
+| `unreachable` | `0x0a` | | trap immediately |
+| `end` | `0x0f` | | end a block, loop, or if |
```

Comment:
@kripken Given the lack of response from @sunfishcode to help us understand his perspective I can only speculate he is trying to shoe-horn the AST into a single value stack machine where zero-value nodes yield no stack element when interpreted. Wasm would be restricted to only being able to consume zero-value expressions in elements of a block that discard their result values thus any zero-value expression seen within a block's node stack is a top-level block node and since these are discarded a value need not be pushed on the stack when interpreted. There would a separate `block1` operator that returns the last expression value, and a separate `block0` operator that returns no values, and while they could be two separate operators @sunfishcode might be suggesting they be distinguished by an immediate 0/1 but I think it would be needed on the block_start marker. The break operators could no long accept a zero-value expression as it is no longer possible to represent this so there would need to be a separate `break0` and it would be a validation error to use these with `block1`. A `breakn` accepting a single non-zero-value expression would be used with `block1` discarding all but the first value, and could also be used with `block0` discarding all the expression values. These two break operators might be distinguished using the 0/1 immediate. The `if` operator already has implicit blocks, and @sunfishcode appears to be suggesting enforcing them both as either being `block0` or `block1`. The `select` operator would no longer be able to accept zero-value expressions, but this does not seem a big constraint.

It would seem to still be necessary to push a zero-value node on the stack when validating or decoding back to an AST, but perhaps once validated the code might be interpreted literally without pushing a value on the stack for a zero-value node.

Scaling to multiple-values might relax `block1` to pass through expression values with one or more values in which case an interpreter needs to deal with the value count anyway so there seems to be little gain restricting the zero-value case?

Also I see mention of plans for wasm to support more general types, I presume a struct that could be empty or have one slot or two slots etc, just as a zero-value or one-value or two-value etc expression would have, so an interpreted needs to deal with these variable sized stack elements anyway and would need to allocate variable size stack elements and track their types. In this context the zero-value restriction seems out of place to me.

Perhaps the intention is that `block1` consume a single value from the fall-through expression and the break expressions and only pass on this single value, but it could still accept a multiple value expression. Alternatively, a `block2` `block3` etc might be added to pass on more than one value, otherwise it would require multiple value expressions to be passed around in local variables and the loss of efficient coding would be the same as if blocks and `if` did not return any values.

CIL failed to plan ahead for multiple-value support and look where it has lead https://github.com/dotnet/roslyn/issues/347 using struct to represent them which is less efficient from an AST perspective. For example the first value can not be trivially consumed discarding the rest etc.

I think people need to give up on trying to shoe horn wasm into an interpreted single value stack machine (JVM style), and to focus on the efficient single pass validation and SSA transform after which it is represented by single value typed definitions. Actually I wonder if expressions were removed from wasm and the SSA transform were followed by a copy-propagation pass if the same SSA code would be obtained? If so then the AST might be view as just a efficient way to represent this.